model/map/src/main/java/org/keycloak/models/map/group/MapGroupProvider.java

```diff
@@ -87,43 +81,47 @@ public GroupModel getGroupById(RealmModel realm, String id) {
             return null;
         }
         
-        MapGroupEntity entity = tx.get(uid, groupStore::get);
-        return (entity == null || ! entityRealmFilter(realm).test(entity))
+        MapGroupEntity entity = tx.read(uid);
+        String realmId = realm.getId();
+        return (entity == null || ! Objects.equals(realmId, entity.getRealmId()))
                 ? null
                 : entityToAdapterFunc(realm).apply(entity);
     }
 
-    private Stream<MapGroupEntity> getNotRemovedUpdatedGroupsStream() {
-        Stream<MapGroupEntity> updatedAndNotRemovedGroupsStream = groupStore.entrySet().stream()
-                .map(tx::getUpdated)    // If the group has been removed, tx.get will return null, otherwise it will return me.getValue()
-                .filter(Objects::nonNull);
-        return Stream.concat(tx.createdValuesStream(), updatedAndNotRemovedGroupsStream);
-    }
-
-    private Stream<MapGroupEntity> getUnsortedGroupEntitiesStream(RealmModel realm) {
-        return getNotRemovedUpdatedGroupsStream()
-                .filter(entityRealmFilter(realm));
-    }
-
     @Override
     public Stream<GroupModel> getGroupsStream(RealmModel realm) {
+        return getGroupsStreamInternal(realm, null);
+    }
+
+    private Stream<GroupModel> getGroupsStreamInternal(RealmModel realm, UnaryOperator<ModelCriteriaBuilder<GroupModel>> modifier) {
         LOG.tracef("getGroupsStream(%s)%s", realm, getShortStackTrace());
-        return getUnsortedGroupEntitiesStream(realm)
+        ModelCriteriaBuilder<GroupModel> mcb = groupStore.createCriteriaBuilder()
+          .compare(SearchableFields.REALM_ID, Operator.EQ, realm.getId());
```

Comment:
This check is now performed in `tx.getUpdatedNotRemoved()` which is the only place where `null` may appear. The set of returned values that were not touched in this transaction should never be `null` (since it would not satisfy the `mcb` filter in the first place).

Reply:
I mean `realm.getId()` will throw NPE when realm is `null`. However, I am not sure this makes sense to add now. We are not checking many other parameters. Maybe we should add it to cleanup task to either document that NPE may be thrown when null parameter is passed or do some checking everywhere. 

I just wanted to point it out as this is change of behaviour in this PR. Before there was no user/group/role returned as ALWAYS_FALSE filter was used when realm was `null`.

Reply:
I agree, this should have been there. Added https://issues.redhat.com/browse/KEYCLOAK-16872